plugins/sinks/shield/sink.go

```diff
@@ -0,0 +1,198 @@
+
+import (
+	"bytes"
+	"context"
+	_ "embed"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/MakeNowJust/heredoc"
+	"github.com/odpf/meteor/models"
+	assetsv1beta2 "github.com/odpf/meteor/models/odpf/assets/v1beta2"
+	"github.com/odpf/meteor/plugins"
+	"github.com/odpf/meteor/registry"
+	"github.com/odpf/salt/log"
+	"github.com/pkg/errors"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/types/known/anypb"
+)
+
+//go:embed README.md
+var summary string
+
+type Config struct {
+	Host    string            `mapstructure:"host" validate:"required"`
+	Headers map[string]string `mapstructure:"headers"`
+}
+
+var info = plugins.Info{
+	Description: "Send user information to shield http service",
+	Summary:     summary,
+	Tags:        []string{"http", "sink"},
+	SampleConfig: heredoc.Doc(`
+	# The hostname of the shield service
+	host: https://shield.com
+	# Additional HTTP headers send to shield, multiple headers value are separated by a comma
+	headers:
+	  X-Shield-Email: meteor@odpf.io
+      X-Other-Header: value1, value2
+	`),
+}
+
+type httpClient interface {
+	Do(*http.Request) (*http.Response, error)
+}
+
+type Sink struct {
+	plugins.BasePlugin
+	client httpClient
+	config Config
+	logger log.Logger
+}
+
+func New(c httpClient, logger log.Logger) plugins.Syncer {
+	s := &Sink{
+		logger: logger,
+		client: c,
+	}
+	s.BasePlugin = plugins.NewBasePlugin(info, &s.config)
+
+	return s
+}
+
+func (s *Sink) Init(ctx context.Context, config plugins.Config) error {
+	if err := s.BasePlugin.Init(ctx, config); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *Sink) Sink(ctx context.Context, batch []models.Record) error {
+	for _, record := range batch {
+		metadata := record.Data()
+		s.logger.Info("sinking record to shield", "record", metadata.GetUrn())
+
+		shieldPayload, err := s.buildShieldPayload(metadata)
+		if err != nil {
+			return errors.Wrap(err, "failed to build shield payload")
+		}
+		if err = s.send(ctx, shieldPayload); err != nil {
+			return errors.Wrap(err, "error sending data")
+		}
+
+		s.logger.Info("successfully sinked record to shield", "record", metadata.Name)
+	}
+
+	return nil
+}
+
+func (s *Sink) Close() (err error) { return }
+
+func (s *Sink) send(ctx context.Context, record RequestPayload) error {
+	payloadBytes, err := json.Marshal(record)
+	if err != nil {
+		return err
+	}
+
+	// send request
+	url := fmt.Sprintf("%s/admin/v1beta1/users/%s", url.PathEscape(s.config.Host), url.PathEscape(record.Email))
+	req, err := http.NewRequestWithContext(ctx, http.MethodPut, url, bytes.NewBuffer(payloadBytes))
+	if err != nil {
+		return err
+	}
+
+	for hdrKey, hdrVal := range s.config.Headers {
+		hdrVals := strings.Split(hdrVal, ",")
+		for _, val := range hdrVals {
+			val = strings.TrimSpace(val)
+			req.Header.Add(hdrKey, val)
+		}
+	}
+
+	res, err := s.client.Do(req)
+	if err != nil {
+		return err
+	}
+	if res.StatusCode == 200 {
```

Comment:
Can we close the response here? Context - https://stackoverflow.com/questions/18598780/is-resp-body-close-necessary-if-we-dont-read-anything-from-the-body/18601625#18601625

You could use the following helper:

```go
// DrainBody drains and closes the response body to avoid the following
// gotcha:
// http://devs.cloudimmunity.com/gotchas-and-common-mistakes-in-go-golang/index.html#close_http_resp_body
func DrainBody(resp *http.Response) {
	if resp == nil {
		return
	}

	_, _ = io.Copy(ioutil.Discard, resp.Body)
	_ = resp.Body.Close()
}
```

Or as suggested by @spy16 in the discussion thread on slack:

```go
type withCleanup struct{ io.ReadCloser }

func (irc *withCleanup) Close() error {
	if irc.ReadCloser == nil {
		return nil
	}
	_, _ = io.Copy(ioutil.Discard, irc.ReadCloser)
	err := irc.ReadCloser.Close()
	irc.ReadCloser = nil
	return err
}
```

> then right after i get response from client, i do `resp.Body = &withCleanup{ resp.Body }` (so normal `defer resp.Body.Close()` ends up doing the proper cleanup sequence)

Article on this gotcha and more: https://devs.cloudimmunity.com/gotchas-and-common-mistakes-in-go-golang/index.html#close_http_resp_body

Reply:
if we have a generic client wrapper, then i would suggest using `withCleanup` approach since the `resp.Body.Close()` would be happening somewhere else..

but if we are directly using the client, then we `defer DrainBody(resp)` would be the way to go 👍